app/screens/image_preview/downloader.ios.js

```diff
@@ -0,0 +1,287 @@
+import Icon from 'react-native-vector-icons/Ionicons';
+
+import {Client4} from 'mattermost-redux/client';
+
+import FormattedText from 'app/components/formatted_text';
+import {emptyFunction} from 'app/utils/general';
+
+const {View: AnimatedView} = Animated;
+const {height: deviceHeight, width: deviceWidth} = Dimensions.get('window');
+
+export default class Downloader extends PureComponent {
+    static propTypes = {
+        file: PropTypes.object.isRequired,
+        onDownloadCancel: PropTypes.func,
+        onDownloadSuccess: PropTypes.func,
+        show: PropTypes.bool
+    };
+
+    static defaultProps = {
+        onCancelPress: emptyFunction,
+        onDownloadSuccess: emptyFunction,
+        show: false
+    };
+
+    state = {
+        downloaderTop: new Animated.Value(deviceHeight),
+        progress: 0
+    }
+
+    componentWillReceiveProps(nextProps) {
+        if (!this.props.show && nextProps.show) {
+            this.toggleDownloader();
+            this.setState({
+                didCancel: false,
+                progress: 0
+            });
+        } else if (!nextProps.show && this.props.show) {
+            this.toggleDownloader(false);
+        }
+    }
+
+    toggleDownloader = (show = true) => {
+        const top = show ? (deviceHeight / 2) - 100 : deviceHeight;
+
+        Animated.spring(this.state.downloaderTop, {
+            toValue: top,
+            tension: 8,
+            friction: 5
+        }).start(() => {
+            if (show) {
+                this.startDownload();
+            }
+        });
+    }
+
+    startDownload = async () => {
+        try {
+            const {file} = this.props;
+            const imageUrl = Client4.getFileUrl(file.id);
+
+            this.downloadTask = RNFetchBlob.config({
+                fileCache: true,
+                appendExt: file.extension
+            }).fetch('GET', imageUrl).progress((received, total) => {
+                const progress = (received / total) * 100;
+                this.setState({
+                    progress
+                });
+            });
+
+            const res = await this.downloadTask;
+            const path = res.path();
+            const newPath = await CameraRoll.saveToCameraRoll(path);
+
+            this.setState({
+                progress: 100
+            }, () => {
+                // need to wait a bit for the progress circle UI to update to the give progress
+                setTimeout(async () => {
+                    try {
+                        // handles the case of a late cancellation by the user
+                        // and ensures that we remove the file if they did cancel late.
+                        if (this.state.didCancel) {
+                            // TODO: There's issue with deleting files from the cameraRoll on iOS here https://github.com/wkh237/react-native-fetch-blob/issues/479
+                            // This only occurs if the user cancels when the download is around 80% or more
```

Comment:
what will happen then? the file will stay there in a corrupt state?

Reply:
No the file is actually fully downloaded and can be used for attachments. Since we use the animated progress circle the file actually downloads before the progress completes since it runs the animation all the way to 100. We can keep an eye on that issue and update the lib when it's fixed.